[WIP] Introduce the notion of builtin templates.

martijnvg opened the pull request:
Today the `TemplateRegisty` classes are responsible for installing templates and ilm policies. There are a few downside of using `TemplateRegisty`:
* A log entry is printed for each resource that is installed, which at this point in time really bloats the logs during startup.
* Each `TemplateRegisty` instance is also a cluster state listener.
* If cluster state updates aren't batched this can lead to many cluster state updates during startup.

Essentially each `TemplateRegisty` is responsible for each own templates and uses the put template APIs to install templates. This makes making templates installation difficult to optimise.

This PR adds `BuiltinTemplatePlugin` interface that plugins can implement. A new internal component is added that installs templates via elected master node only. All registered templates are installed always in a single cluster state update and a single log line is printed for each resource type (component templates, composable index templates etc.).

This is currently just a WIP and tests probably fail. This WIP only addresses the installation of builtin component and composable index templates. Support for installing builtin legacy templates (used by monitoring) still needs to be added. Also ILM policies aren't addressed, but this is a bit trickier since no notion of ilm exists in core Elasticsearch. But this can be addressed in a  followup change. In the current state, this PR should already cleanup ES logs during startup and reduce startup time.

Template related startup log with this change:

```
[2022-04-21T21:24:27,311][INFO ][o.e.c.m.BuiltinTemplateManager] [runTask-0] installed component templates [data-streams-mappings, synthetics-mappings, metrics-settings, metrics-mappings, synthetics-settings, logs-mappings, logs-settings, .deprecation-indexing-mappings, .deprecation-indexing-settings]
[2022-04-21T21:24:28,066][INFO ][o.e.c.m.BuiltinTemplateManager] [runTask-0] installed composable index templates [.monitoring-beats-mb, .monitoring-es-mb, .monitoring-ent-search-mb, .monitoring-logstash-mb, .monitoring-kibana-mb, .ml-notifications-000002, .ml-anomalies-, .ml-state, .ml-stats, logs, synthetics, metrics, .watch-history-16, .slm-history, ilm-history, .deprecation-indexing-template]
[2022-04-21T21:24:28,211][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding template [.monitoring-beats] for index patterns [.monitoring-beats-7-*]
[2022-04-21T21:24:28,313][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding template [.monitoring-logstash] for index patterns [.monitoring-logstash-7-*]
[2022-04-21T21:24:28,409][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding template [.monitoring-alerts-7] for index patterns [.monitoring-alerts-7]
[2022-04-21T21:24:28,490][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding template [.monitoring-kibana] for index patterns [.monitoring-kibana-7-*]
[2022-04-21T21:24:28,618][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding template [.monitoring-es] for index patterns [.monitoring-es-7-*]
[2022-04-21T21:24:28,706][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [30-days-default]
[2022-04-21T21:24:28,832][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [logs]
[2022-04-21T21:24:28,923][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [synthetics]
[2022-04-21T21:24:29,033][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [365-days-default]
[2022-04-21T21:24:29,148][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [180-days-default]
[2022-04-21T21:24:29,247][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [.monitoring-8-ilm-policy]
[2022-04-21T21:24:29,355][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [90-days-default]
[2022-04-21T21:24:29,477][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [metrics]
[2022-04-21T21:24:29,581][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [7-days-default]
[2022-04-21T21:24:29,707][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [ml-size-based-ilm-policy]
[2022-04-21T21:24:29,835][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [.deprecation-indexing-ilm-policy]
[2022-04-21T21:24:29,917][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [watch-history-ilm-policy-16]
[2022-04-21T21:24:30,029][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [ilm-history-ilm-policy]
[2022-04-21T21:24:30,105][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [slm-history-ilm-policy]
[2022-04-21T21:24:30,248][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [.fleet-actions-results-ilm-policy]
```

(this should be much shorter when other plugins use `BuiltinTemplatePlugin` and when legacy templates are supported too)

and without:

```
[2022-04-20T23:38:11,866][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding index template [.monitoring-logstash-mb] for index patterns [.monitoring-logstash-8-*]
[2022-04-20T23:38:11,988][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding index template [.ml-state] for index patterns [.ml-state*]
[2022-04-20T23:38:12,257][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding index template [.monitoring-es-mb] for index patterns [.monitoring-es-8-*]
[2022-04-20T23:38:12,391][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding index template [.ml-stats] for index patterns [.ml-stats-*]
[2022-04-20T23:38:12,501][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding index template [.ml-notifications-000002] for index patterns [.ml-notifications-000002]
[2022-04-20T23:38:12,624][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding index template [.ml-anomalies-] for index patterns [.ml-anomalies-*]
[2022-04-20T23:38:12,756][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding index template [.monitoring-ent-search-mb] for index patterns [.monitoring-ent-search-8-*]
[2022-04-20T23:38:12,910][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding index template [.monitoring-kibana-mb] for index patterns [.monitoring-kibana-8-*]
[2022-04-20T23:38:13,124][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding index template [.monitoring-beats-mb] for index patterns [.monitoring-beats-8-*]
[2022-04-20T23:38:13,246][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding template [.monitoring-alerts-7] for index patterns [.monitoring-alerts-7]
[2022-04-20T23:38:13,348][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding template [.monitoring-logstash] for index patterns [.monitoring-logstash-7-*]
[2022-04-20T23:38:13,456][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding template [.monitoring-beats] for index patterns [.monitoring-beats-7-*]
[2022-04-20T23:38:13,548][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding template [.monitoring-kibana] for index patterns [.monitoring-kibana-7-*]
[2022-04-20T23:38:13,658][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding component template [data-streams-mappings]
[2022-04-20T23:38:13,767][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding component template [synthetics-mappings]
[2022-04-20T23:38:13,849][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding component template [metrics-settings]
[2022-04-20T23:38:14,001][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding component template [metrics-mappings]
[2022-04-20T23:38:14,082][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding component template [logs-settings]
[2022-04-20T23:38:14,178][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding component template [synthetics-settings]
[2022-04-20T23:38:14,321][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding component template [logs-mappings]
[2022-04-20T23:38:14,419][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding template [.monitoring-es] for index patterns [.monitoring-es-7-*]
[2022-04-20T23:38:14,690][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding index template [.watch-history-16] for index patterns [.watcher-history-16*]
[2022-04-20T23:38:14,806][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding index template [ilm-history] for index patterns [ilm-history-5*]
[2022-04-20T23:38:14,896][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding component template [.deprecation-indexing-mappings]
[2022-04-20T23:38:14,973][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding component template [.deprecation-indexing-settings]
[2022-04-20T23:38:15,213][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding index template [.slm-history] for index patterns [.slm-history-5*]
[2022-04-20T23:38:15,408][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding index template [metrics] for index patterns [metrics-*-*]
[2022-04-20T23:38:15,513][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding index template [synthetics] for index patterns [synthetics-*-*]
[2022-04-20T23:38:15,627][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding index template [logs] for index patterns [logs-*-*]
[2022-04-20T23:38:15,731][INFO ][o.e.c.m.MetadataIndexTemplateService] [runTask-0] adding index template [.deprecation-indexing-template] for index patterns [.logs-deprecation.*]
[2022-04-20T23:38:15,861][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [.monitoring-8-ilm-policy]
[2022-04-20T23:38:15,968][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [ml-size-based-ilm-policy]
[2022-04-20T23:38:16,042][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [metrics]
[2022-04-20T23:38:16,123][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [30-days-default]
[2022-04-20T23:38:16,209][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [90-days-default]
[2022-04-20T23:38:16,306][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [180-days-default]
[2022-04-20T23:38:16,402][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [logs]
[2022-04-20T23:38:16,477][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [synthetics]
[2022-04-20T23:38:16,550][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [365-days-default]
[2022-04-20T23:38:16,642][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [7-days-default]
[2022-04-20T23:38:16,745][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [watch-history-ilm-policy-16]
[2022-04-20T23:38:16,868][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [ilm-history-ilm-policy]
[2022-04-20T23:38:16,947][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [.fleet-actions-results-ilm-policy]
[2022-04-20T23:38:17,150][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [runTask-0] adding index lifecycle policy [slm-history-ilm-policy]
```

martijnvg committed:
init

martijnvg committed:
spotless

martijnvg committed:
spotless

martijnvg committed:
spotless

martijnvg committed:
fixed tests and compile error

martijnvg committed:
Merge remote-tracking branch 'es/master' into builtin_templates

martijnvg committed:
cut other xpack modules over to use BuiltinTemplatePlugin
and tweak login in BuiltinTemplateManager.

martijnvg committed:
Merge remote-tracking branch 'es/master' into builtin_templates

martijnvg committed:
spotless

martijnvg committed:
fixed various compile, checkstyle and spotless errors

rjernst commented:
I like the idea! Could we consider using SPI instead of a new plugin interface? It seems especially doable in this case since (at least at a glance from looking at the PR) these templates are all static anyways, so there is no state the SPI impl classes would need.

martijnvg committed:
Made builtin templates pluggable via `TemplateBundle` that are loaded via SPI instead of `BuiltinTemplatePlugin` plugin extension.

martijnvg commented:
Thanks for checking this out Ryan!

> Could we consider using SPI instead of a new plugin interface? 

I think this works out. I implemented this via: 717cf30cad32628c34f9288a73f92d2d32c8d1aa

I added a method to `PluginService` (`loadExtension(...)` method) to allow plugins to define `TemplateBundle` via SPI. This interface is defined in server module, and I didn't see another way to make this work. (`ExtensiblePlugin` is defining extension points in plugins and not in server, right?). But maybe I'm wrong here and there is a better way.

There are also 2 TODOs regarding plugins that allow their templates to be enabled / disabled at runtime. I think there is way this can be implemented. I think I can add a method to `TemplateBundle` where plugins can define which `Setting` the `TemplateBundleManager` should check for whether templates have been disabled and if so skip checking whether templates need to be updated/installed.

martijnvg committed:
variable rename

martijnvg committed:
removed unused IndexTemplateRegistry code and
removed SnapshotLifecycleTemplateRegistry#validate(...) method,
since it wasn't used in production code.

martijnvg committed:
allow registering a setting that enables/disables templates

martijnvg committed:
add slmHistoryEnabled to LifecycleTemplateBundle

dakrone commented:
We discussed this a bunch and I think we could go forward with something like this, but going to close this one for now until we pick the work back up.
